packages/datadog-instrumentations/src/http.js

```diff
@@ -0,0 +1,210 @@
+          errorClientCh.publish(arg)
+
+          throw err
+        } finally {
+          endClientCh.publish(undefined)
+        }
+      }
+      bindEmit(req)
+      return req
+    }
+  }
+
+  function normalizeArgs (inputURL, inputOptions, cb) {
+    inputURL = normalizeOptions(inputURL)
+
+    const [callback, inputOptionsNormalized] = normalizeCallback(inputOptions, cb, inputURL)
+    const options = combineOptions(inputURL, inputOptionsNormalized)
+    normalizeHeaders(options)
+    const uri = url.format(options)
+
+    return { uri, options, callback }
+  }
+
+  function combineOptions (inputURL, inputOptions) {
+    if (typeof inputOptions === 'object') {
+      return Object.assign(inputURL || {}, inputOptions)
+    } else {
+      return inputURL
+    }
+  }
+  function normalizeHeaders (options) {
+    options.headers = options.headers || {}
+  }
+
+  function normalizeCallback (inputOptions, callback, inputURL) {
+    if (typeof inputOptions === 'function') {
+      return [inputOptions, inputURL || {}]
+    } else {
+      return [callback, inputOptions]
+    }
+  }
+
+  function normalizeOptions (inputURL) {
+    if (typeof inputURL === 'string') {
+      try {
+        return urlToOptions(new url.URL(inputURL))
+      } catch (e) {
+        return url.parse(inputURL)
+      }
+    } else if (inputURL instanceof url.URL) {
+      return urlToOptions(inputURL)
+    } else {
+      return inputURL
+    }
+  }
+
+  function urlToOptions (url) {
+    const agent = url.agent || http.globalAgent
+    const options = {
+      protocol: url.protocol || agent.protocol,
+      hostname: typeof url.hostname === 'string' && url.hostname.startsWith('[')
+        ? url.hostname.slice(1, -1)
+        : url.hostname ||
+        url.host ||
+        'localhost',
+      hash: url.hash,
+      search: url.search,
+      pathname: url.pathname,
+      path: `${url.pathname || ''}${url.search || ''}`,
+      href: url.href
+    }
+    if (url.port !== '') {
+      options.port = Number(url.port)
+    }
+    if (url.username || url.password) {
+      options.auth = `${url.username}:${url.password}`
+    }
+    return options
+  }
+}
+
+addHook({ name: 'https' }, http => {
+  patch.call(this, http, 'request')
+  patch.call(this, http, 'get')
+
+  shimmer.wrap(http.Server.prototype, 'emit', wrapEmit)
+
+  return http
+})
+
+addHook({ name: 'http' }, http => {
+  patch.call(this, http, 'request')
+  patch.call(this, http, 'get')
+
+  shimmer.wrap(http.Server.prototype, 'emit', wrapEmit)
+
+  if (http.ServerResponse) { // not present on https
+    bindEmit(http.ServerResponse.prototype)
+  }
+  return http
+})
+
+function wrapEmit (emit) {
+  return function (eventName, req, res) {
+    if (!startServerCh.hasSubscribers) {
+      return emit.apply(this, arguments)
+    }
+
+    if (eventName === 'request') {
+      return web.web.instrument(startServerCh, req, res, () => {
+        if (incomingHttpRequestStart.hasSubscribers) {
```

Comment:
I'll be sure to remove those once AppSec transitions to using the new instrumentation channels 👍